### DISCUSSION TITLE
Pycalphad rearranges constituent array for paramater when passed to database

### DISCUSSION THREAD

**Discussion by wahab2604** [pycalphad/pycalphad#314] [Q&A] [answered]
I have 2 liquid interaction parameters in my TDB file whose constituents get passed to pycalphad and plotted in the incorrect order. 

E.g PARAMETER L (SIO2LIQ, CSLIQ) gets changed to PARAMETER L (CSLIQ, SIO2LIQ)
This results in the phase diagram not being outputted correctly. I discovered this after running a quick calc in ESPEI and looking at the output tdb (this is even the case when the parameter is excluded from the calculation) 

I think its do with some alphabetical sorting of the species dict because when I rename the species and change the consitutent array in order of alphabetical importance then it works
i.e 
PARAMETER L (BSIO2LIQ, CSLIQ) doesnt change when passed to pycalphad and plots the correct phase diagram 

Is there a way to force pycalphad to keep the desired consituent array regardless of its name?

**Comment by bocklund** [comment, marked as answer]
Hi @wahab2604, thank you for your question. pycalphad sorts the species on each sublattice intentionally when reading a TDB ([see the code here](https://github.com/pycalphad/pycalphad/blob/50389503f126ea6f9e40947c215bd857bcfbd9de/pycalphad/io/tdb.py#L365) and also #17) because the (x<sub>A</sub> - x<sub>B</sub> )<sup>ν</sup> term in the Redlich-Kister polynomial means that (if left unsorted) `L(PHASE,A,B;<odd order interaction>) = -L(PHASE,B,A;<odd order interaction>)`.

The constituents get sorted alphabetically when pycalphad reads the TDB file, so that is why they are sorted when pycalphad (and ESPEI) writes the TDB file. Since the other software packages also do this alphabetic sorting, storing and writing out the parameters with constituents alphabetically sorted ensures that the parameters in the TDB translate to the excess Gibbs energies correctly. In that spirit, we don't currently provide an option to turn off the alphabetic sorting.

**Comment by wahab2604** [comment]
Hi , 
Thank you for your reply, I understand why that sorting is done now
